src/util/address.rs

```diff
@@ -143,11 +154,167 @@ impl FromStr for AddressType {
+            0 => WitnessVersion::V0,
+            1 => WitnessVersion::V1,
+            2 => WitnessVersion::V2,
+            3 => WitnessVersion::V3,
+            4 => WitnessVersion::V4,
+            5 => WitnessVersion::V5,
+            6 => WitnessVersion::V6,
+            7 => WitnessVersion::V7,
+            8 => WitnessVersion::V8,
+            9 => WitnessVersion::V9,
+            10 => WitnessVersion::V10,
+            11 => WitnessVersion::V11,
+            12 => WitnessVersion::V12,
+            13 => WitnessVersion::V13,
+            14 => WitnessVersion::V14,
+            15 => WitnessVersion::V15,
+            16 => WitnessVersion::V16,
+            wrong => Err(Error::InvalidWitnessVersion(wrong))?,
+        })
+    }
+
+    /// Takes bitcoin Script opcode and returns either corresponding version of
+    /// the Witness program (for opcodes in range of `OP_0`..`OP_16`) or
+    /// [`WitnessVersionError`] error for the rest of opcodes
+    pub fn from_opcode(opcode: opcodes::All) -> Result<Self, Error> {
+        match opcode.into_u8() {
+            0 => Ok(WitnessVersion::V0),
+            ver if ver >= opcodes::all::OP_PUSHNUM_1.into_u8() && ver <= opcodes::all::OP_PUSHNUM_16.into_u8() =>
+                WitnessVersion::from_ver_u8(ver - opcodes::all::OP_PUSHNUM_1.into_u8() + 1),
+            _ => Err(Error::UnparsableWitnessVersion)
+        }
+    }
+
+    /// Takes bitcoin Script instruction (parsed opcode) and returns either
+    /// corresponding version of the Witness program (for push-num instructions)
+    /// or [`WitnessVersionError`] error for the rest of opcodes
+    pub fn from_instruction(instruction: Instruction) -> Result<Self, Error> {
+        match instruction {
+            Instruction::Op(op) => WitnessVersion::from_opcode(op),
+            Instruction::PushBytes(bytes) if bytes.len() == 0 => Ok(WitnessVersion::V0),
+            Instruction::PushBytes(_) => Err(Error::UnparsableWitnessVersion),
+        }
+    }
+
+    /// Returns integer version number representation for a given [`WitnessVersion`] value.
+    /// NB: this is not the same as an integer representation of the opcode signifying witness
+    /// version in bitcoin script. Thus, there is no function to directly convert witness version
+    /// into a byte since the conversion requires context (bitcoin script or just a version number)
+    pub fn into_ver_u8(self) -> u8 {
+        self as u8
+    }
+
+    /// Determine the checksum variant. See BIP-0350 for specification.
+    pub fn bech32_variant(&self) -> bech32::Variant {
+        match self {
+            WitnessVersion::V0 => bech32::Variant::Bech32,
+            _ => bech32::Variant::Bech32m,
+        }
+    }
+}
+
+impl From<WitnessVersion> for ::bech32::u5 {
+    /// Converts `WitnessVersion` instance into corresponding Bech32 u5-value
+    fn from(ver: WitnessVersion) -> Self {
+        ::bech32::u5::try_from_u8(ver.into_ver_u8()).expect("WitnessVersion must be 0..=16")
+    }
+}
+
```

Comment:
Maybe conersion to `u8` would be useful as well?

Reply:
We can't have it since it should be a failing conversion and with rustc 0.29 we do not have `TryFrom`. To convert `WitnessVersion` to `u8` we already have a dedicated function